File: _3_Learning/_02_Algorithms/Week2/Deque.py
class Item():
    def __init__(self, item):
        self.item = item
        self.prev = None
        self.next = None


class ListIterator():
    def __init__(self, first):
        self.current = first

    def hasNext(self):
        return self.current != None

    def remove(self):
        raise UnsupportedOperationException

    def next(self):
        if not self.hasNext():
            raise NoSuchElementException
        item = self.current.item
        self.current = self.current.next
        return item


class Deque:
    def __init__(self):
        self.first = None
        self.last = None
        self.n = 0

    def isEmpty(self):
        return self.n == 0

    def size(self):
        return self.n

    def addFirst(self, item):
        # Add to Front
        if None == item:
            raise IllegalArgumentException
        oldfirst = self.first
        self.first = Item(item)
        self.first.next = oldfirst
        if self.isEmpty():
            self.last = self.first
        else:
            oldfirst.prev = self.first
        self.n += 1

    def addLast(self, item):
        # Add to Back
        if None == item:
            raise IllegalArgumentException
        oldlast = self.last
        self.last = Item(item)
        self.last.prev = oldlast
        if self.isEmpty():
            self.first = self.last
        else:
            oldlast.next = self.last
        self.n+=1

    def removeFirst(self):
        # pop front
        if self.isEmpty():
            raise NoSuchElementException
        item = self.first.item
        newfirst = self.first.next
        if newfirst:
            newfirst.prev = None
        else:
            self.last = None
        self.first = newfirst
        self.n -= 1
        return item

    def removeLast(self):
        if self.isEmpty():
            raise NoSuchElementException
        item = self.last.item
        newlast = self.last.prev
        if newlast:
            newlast.next = None
        else:
            self.first=None
        self.last = newlast
        self.n -= 1
        return item

    def iterator(self):
        # return a iter
        return ListIterator(self.first)


class IllegalArgumentException(Exception):
    pass


class NoSuchElementException(Exception):
    pass


class UnsupportedOperationException(Exception):
    pass


if __name__ == '__main__':
    deque = Deque()
    print(deque.size())
    deque.addFirst('item1')
    print(deque.removeLast())
    print(deque.size())
    deque.addFirst('item2')
    deque.addFirst('item3')
    it = deque.iterator()
    while it.hasNext():
        print(it.next())
    print(deque.size())
    print(deque.removeLast())
    print(deque.size())
    deque.addLast('item4')
    print(deque.size())
    print(deque.removeFirst())
    print(deque.removeFirst())

File: _3_Learning/_02_Algorithms/Week2/Deque.java
import java.util.Iterator;
import java.util.NoSuchElementException;

public class Deque<Item> implements Iterable<Item> {
    private Node<Item> first;    // beginning of queue
    private Node<Item> last;     // end of queue
    private int n;               // number of elements on queue

    // helper linked list class
    private static class Node<Item> {
        private Item item;
        private Node<Item> prev;
        private Node<Item> next;
    }

    // construct an empty deque
    public Deque() {
        first = null;
        last = null;
        n = 0;
    }

    // is the deque empty?
    public boolean isEmpty() {
        return n == 0;
    }

    // return the number of items on the deque
    public int size() {
        return n;
    }

    // add the item to the front
    public void addFirst(Item item) {
        if (item == null) throw new IllegalArgumentException();
        Node<Item> oldfirst = first;
        first = new Node<Item>();
        first.item = item;
        first.prev = null;
        first.next = oldfirst;
        if (isEmpty()) last = first;
        else oldfirst.prev = first;
        n++;
    }

    // add the item to the back
    public void addLast(Item item) {
        if (item == null) throw new IllegalArgumentException();
        Node<Item> oldlast = last;
        last = new Node<Item>();
        last.item = item;
        last.next = null;
        last.prev = oldlast;
        if (isEmpty()) first = last;
        else oldlast.next = last;
        n++;
    }

    // remove and return the item from the front
    public Item removeFirst() {
        if (isEmpty()) throw new NoSuchElementException();
        Item item = first.item;
        Node<Item> newfirst = first.next;
        if (null != newfirst) newfirst.prev = null;
        else last = null;
        first = newfirst;
        n--;
        return item;
    }


    // remove and return the item from the back
    public Item removeLast() {
        if (isEmpty()) throw new NoSuchElementException();
        Item item = last.item;
        Node<Item> newlast = last.prev;
        if (null != newlast) newlast.next = null;
        else first = null;
        last = newlast;
        n--;
        return item;
    }

    // return an iterator over items in order from front to back
    public Iterator<Item> iterator() {
        return new ListIterator(first);
    }

    private class ListIterator implements Iterator<Item> {
        private Node<Item> current;

        public ListIterator(Node<Item> first) {
            current = first;
        }

        public boolean hasNext() {
            return current != null;
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }

        public Item next() {
            if (!hasNext()) throw new NoSuchElementException();
            Item item = current.item;
            current = current.next;
            return item;
        }
    }

    // unit testing (required)
    public static void main(String[] args) {
        Deque<String> deque = new Deque<String>();
        System.out.println(deque.isEmpty());
        // //        input = "to be or not to - be - - that - - - is";
        // while (!StdIn.isEmpty()) {
        //     String item = StdIn.readString();
        //     if (!item.equals("-"))
        //         queue.enqueue(item);
        //     else if (!queue.isEmpty())
        //         StdOut.print(queue.dequeue() + " ");
        // }
        // StdOut.println("(" + queue.size() + " left on queue)");
    }

}
